### DISCUSSION TITLE
chain rules for different upstream proxies based on url filters

### DISCUSSION THREAD

**Discussion by zioalex** [mitmproxy/mitmproxy#5173] [Q&A] [answered]
Hi All,
I was trying to understand if I can use mitmproxy to forward requests to 2 different upstream proxies based on simple url filters.

I can easily setup the http_proxy variable so a normal setup would work but I didn't find it the documented.
Is it something I can do?
Thanks
Alessandro

**Comment by Prinzhorn** [comment, marked as answer]
Not an expert on this topic, but you can try setting `Server.via` inside the `server_connect` hook if `Server.address` (host, port) is enough for you to determine the upstream proxy.

https://docs.mitmproxy.org/stable/api/events.html#ConnectionEvents.server_connect
https://docs.mitmproxy.org/stable/api/mitmproxy/connection.html#Server

**Reply by mhils** [reply]
Adding to @Prinzhorn's answer: https://github.com/mitmproxy/mitmproxy/blob/main/examples/contrib/change_upstream_proxy.py :)

**Comment by zioalex** [comment]
Hi all,
thanks.
I think that now I understood how to change the upstream proxy but not sure how can I filter based on the called url.

I think that the [filters ](https://docs.mitmproxy.org/stable/concepts-filters/) could be used but not sure how to hook them in the change_upstream_proxy.py example.

**Reply by Prinzhorn** [reply]
Use the power of `if` / `else` https://github.com/mitmproxy/mitmproxy/blob/3387dac679ee92ef127097c98455b4f229f4a307/examples/contrib/change_upstream_proxy.py#L21-L25

**Reply by zioalex** [reply]
Yep, I see now that the class flow.request.host contains exactly the the target host name.
I thought that I had to use https://docs.mitmproxy.org/stable/api/mitmproxy/http.html#Request but may be they are the same thing.
I'll try it soon.
Thanks for the support!
By the way this piece of SW is amazing...It looks like you can what ever you want around proxy and tcp....Really impressed.

**Reply by Prinzhorn** [reply]
There are full API docs available, please read them. `flow.request` has a lot more properties https://docs.mitmproxy.org/stable/api/mitmproxy/http.html#Request

**Reply by ErebusST** [reply]
And how to configure the username and password of upstream?
